jina/executors/indexers/vector.py

```diff
@@ -123,78 +130,92 @@ def build_advanced_index(self, vecs: 'np.ndarray'):
 
-        if self.int2ext_key is not None and vecs is not None and vecs.ndim == 2:
-            if self.int2ext_key.shape[0] != vecs.shape[0]:
+    @cached_property
+    def int2ext_id(self) -> Optional['np.ndarray']:
+        """Convert internal ids (0,1,2,3,4,...) to external ids (random index) """
+        if self.key_bytes and self.key_dtype:
+            r = np.frombuffer(self.key_bytes, dtype=self.key_dtype)
+            if r.shape[0] == self.size == self.raw_ndarray.shape[0]:
+                return r
+            else:
                 self.logger.error(
-                    f'the size of the keys and vectors are inconsistent ({self.int2ext_key.shape[0]} != {vecs.shape[0]}), '
+                    f'the size of the keys and vectors are inconsistent '
+                    f'({r.shape[0]}, {self._size}, {self.raw_ndarray.shape[0]}), '
                     f'did you write to this index twice? or did you forget to save indexer?')
-                return None
-            if vecs.shape[0] == 0:
-                self.logger.warning(f'an empty index is loaded')
 
-            self.ext2int_key = {k: idx for idx, k in enumerate(self.int2ext_key)}
-            return vecs
-        else:
-            return None
+    @cached_property
+    def ext2int_id(self) -> Optional[Dict]:
+        """Convert external ids (random index) to internal ids (0,1,2,3,4,...) """
+        if self.int2ext_id is not None:
+            return {k: idx for idx, k in enumerate(self.int2ext_id)}
 
-    def query_by_id(self, ids: Union[List[int], 'np.ndarray'], *args, **kwargs) -> 'np.ndarray':
-        int_ids = np.array([self.ext2int_key[j] for j in ids])
-        return self.raw_ndarray[int_ids]
+
+@lru_cache(maxsize=3)
+def _get_ones(x, y):
+    return np.ones((x, y))
 
 
-def _ext_arrs(A, B):
+def _ext_A(A):
     nA, dim = A.shape
-    A_ext = np.ones((nA, dim * 3))
+    A_ext = _get_ones(nA, dim * 3)
     A_ext[:, dim:2 * dim] = A
     A_ext[:, 2 * dim:] = A ** 2
+    return A_ext
 
-    nB = B.shape[0]
-    B_ext = np.ones((dim * 3, nB))
+
+def _ext_B(B):
+    nB, dim = B.shape
+    B_ext = _get_ones(dim * 3, nB)
     B_ext[:dim] = (B ** 2).T
     B_ext[dim:2 * dim] = -2.0 * B.T
-    return A_ext, B_ext
+    del B
+    return B_ext
 
 
-def _euclidean(A, B):
-    A_ext, B_ext = _ext_arrs(A, B)
+def _euclidean(A_ext, B_ext):
     sqdist = A_ext.dot(B_ext).clip(min=0)
     return np.sqrt(sqdist)
 
 
-def _cosine(A, B):
-    A_ext, B_ext = _ext_arrs(A / np.linalg.norm(A, ord=2, axis=1, keepdims=True),
-                             B / np.linalg.norm(B, ord=2, axis=1, keepdims=True))
-    return A_ext.dot(B_ext).clip(min=0) / 2
+def _norm(A):
+    return A / np.linalg.norm(A, ord=2, axis=1, keepdims=True)
+
+
+def _cosine(A_norm_ext, B_norm_ext):
+    return A_norm_ext.dot(B_norm_ext).clip(min=0) / 2
 
 
 class NumpyIndexer(BaseNumpyIndexer):
     """An exhaustive vector indexers implemented with numpy and scipy. """
 
+    batch_size = 512
```

Comment:
what if we put it as parameter and have it in a `cached_property` so that is not dumped?

Reply:
`batch_size` is part of `metas` and for `metas`, the priority is 

metas in YAML > class attribute > default_jina_config

Reply:
btw our docs on `metas` has been broken for quite a while